### DISCUSSION TITLE
Make all handlers as subclass of CH with single state

### DISCUSSION THREAD

**Discussion by david-shiko** [python-telegram-bot/python-telegram-bot#2853] [Q&A]
I looked at code here https://github.com/python-telegram-bot/python-telegram-bot/blob/855d711630851894f87a6ae6bf7b9546c8c5171e/telegram/ext/_conversationhandler.py#L660
(Most relevant dev version of CH which I found)
(Still in development)

This may be not so convenient because all `check_update` of `*handler` returns 3 values and here is 4.
Technically every handler is a subclass of CH with a single dict state.
So other handlers will always return `-1` as a state (`return state, key, handler, check`)

Even more, `dispatcher` may be presented as CH and every state is similar to a group. 
Eventually, it's going to be some sort of tree where `dispatcher` is a root CH.

Additionally, we can move a `dispatcher.context_types.context.from_update(update, self)` inside `handler_update` because a `state` is only a one of many parameters of CH (or another handler) which may be useful sometimes

P.S. is `context` may be `True` (not `None`) at this [point](https://github.com/python-telegram-bot/python-telegram-bot/blob/ade1529986f5b6d394a65372d6a27045a70725b2/telegram/ext/dispatcher.py#L550) because [above](https://github.com/python-telegram-bot/python-telegram-bot/blob/ade1529986f5b6d394a65372d6a27045a70725b2/telegram/ext/dispatcher.py#L541) he's explicit `None`

**Comment by Bibo-Joshi** [comment]
Hi. TBH I don't really get what the punch line of your post is 🤔 Are you suggesting a code change? If so, could you explain it?
Still, some replies to your comments:

* the `check_update` method of a handler may return *any* object and it will be passed to the `handle_update` method.
* From an inheritance pov, `CH` is a special instance of `Handler`, not the other way around. your pov makes sense, if you assume *every* bot to be some sort of FSM, which is in general not the case
* handler groups have little to do with CH states - only one state at a time can be active, but an update can be processed by handlers from multiple groups
* moving the building of the `context` argument inside `handle_update` would mean that users could no longer set custom attributes on the `context` argument and access them in handlers of later groups - which is a documented feature.

**Reply by david-shiko** [reply]
Just found a new tab "your discussions" in github :)

Sorry for my grammar errors and unclear message.

1. TBH I don't remember what was a sense with num of arguments :)
2. I don't understood your message about FSM :)
3. It's may be a cool feature - represent a CH as tree (of groups of handlers).
4. _"context may be True (not None)"_ - no need `if not context` at [this](
https://github.com/python-telegram-bot/python-telegram-bot/blob/ade1529986f5b6d394a65372d6a27045a70725b2/telegram/ext/dispatcher.py#L550) 
line because context (most likely) are always `None`.
